media.go

```diff
@@ -446,6 +488,17 @@ func (ms *mediaStream[T, U]) Next(ctx context.Context) (T, func(), error) {
 }
 
 func (ms *mediaStream[T, U]) Close(ctx context.Context) error {
+	// We want to create a new span that lasts the scope of this function and no
+	// further, so we have to modify the cancelCtx and then revert it at the end
+	// of the function to what it was before.
+	originalCancelCtx := ms.prodCon.cancelCtx
+	parentSpan := trace.FromContext(ctx)
+	ms.prodCon.cancelCtx = trace.NewContext(ms.prodCon.cancelCtx, parentSpan)
```

Comment:
Why this construct of `StartSpan`? It looks like `FromContext` may return `nil` if there's no span in `ctx`. It may then pass `nil` (assuming `parentSpan == nil`) to `NewContext` which will map the `nil` value to a key. You can test how this would behave by trying `ms.prodCon.cancelCtx = trace.NewContext(ms.prodCon.cancelCtx, nil)`. Ideally we'd create a new span if one doesn't already exists which is what `StartSpan` does.

Reply:
The idea is that there's a span in `ctx` which is the parent of the new span, so we take it and make it `ms.prodCon.cancelCtx`'s current span and then call `StartSpan` on it to create the child. If I'm understanding correctly, your suggestion would be `ms.prodCon.cancelCtx = trace.StartSpan(ms.prodCon.cancelCtx, "gostream::mediastream::Close")`, but then the span from `ctx` isn't the parent. I was assuming that `ctx` would always hold a span since it's called from `ReadMedia()` which itself puts a span in the context, but you're right that it's important to make sure the span isn't nil first. I can make that change